README.md
# react-grid-layout-builder

React-grid-layout-builder offers a React user interface to interact with the [react-grid-layout](https://github.com/STRML/react-grid-layout) configuration. 
Be careful:
* React-grid-layout-builder doesn't add additional features to existing React-grid-layout. Read more about [React-grid-layout features](https://github.com/STRML/react-grid-layout#features) if required.
* React-grid-layout-builder works only on [responsive react-grid-layout](https://github.com/STRML/react-grid-layout#responsive-usage). Any help on other React-grid-layout component is welcome :)

## Motivation

[React-grid-layout](https://github.com/STRML/react-grid-layout) is fun. I wanted to play with it without coding the configuration. So I tried to make a user-friendly editor. Happy to share it now :)

## Demos

1. [Bootstrap editor DEMO](http://damienleroux.github.io/react-grid-layout-builder/demo/bootstrap/index.html)
2. [Material editor DEMO](http://damienleroux.github.io/react-grid-layout-builder/demo/material-ui/index.html)

![reactgridlayoutbuilderdemo](https://cloud.githubusercontent.com/assets/12717418/15196785/5ab09438-17ce-11e6-87dd-11bfc90c7526.gif)


## Installation

### To try it locally:

 - clone this repository
 - go into the cloned folder
 - cmd `npm i`
 - cmd `npm run start`
 - open urls
    - [http://localhost:3100/bootstrap/index.html](http://localhost:3100/bootstrap/index.html) for bootstrap editor
    - [http://localhost:3100/material-ui/index.html](http://localhost:3100/material-ui/index.html) for material-ui editor

### To install as dependency:

Install the React-Grid-Layout-builder [package](https://www.npmjs.com/package/react-grid-layout-builder) package using [npm](https://www.npmjs.com/):

```bash
npm install react-grid-layout-builder
```

Include the following stylesheets in your application:

```
/node_modules/react-grid-layout-builder/css/styles.css
/node_modules/react-grid-layout/css/styles.css
/node_modules/react-resizable/css/styles.css

<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/latest/css/bootstrap.min.css">
<link rel="stylesheet" type="text/css" href="stylesheets/font-awesome/css/font-awesome.min.css">
```

This builder editor is graphically powered by [Bootstrap](http://getbootstrap.com/getting-started/)

Include the following stylesheets in your application:

```
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/latest/css/bootstrap.min.css">
```
Or create your own editor on the model given in [src/bootstrapEditor](https://github.com/damienleroux/react-grid-layout-builder/blob/master/src/bootstrapEditor.js). You'll be able to add fields or remove others
In that case, you'll have to wrap it with `connectReactGridLayoutBuilderToEditor`. Example:

```javascript

//YourOwnBuilderEditor is a copy of src/bootstrapEditor with your own style and your own fields
var ReactGridLayoutBuilder = connectReactGridLayoutBuilderToEditor(YourOwnBuilderEditor);
export default ReactGridLayoutBuilder;
```


This opening dock icon is graphically powered by [Bootstrap](http://fontawesome.io/)

Include the following stylesheets in your application if you use it:

```
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/latest/css/bootstrap.min.css">
```
Or put the editor anywhere else on your app.

## Features

React-grid-layout-builder is to use on the [responsive react-grid-layout](https://github.com/STRML/react-grid-layout#responsive-usage) only and can be used to:

* track any changes when playing with React-grid-layout
* edit number of columns on the fly
* edit breakpoints pixel value on the fly
* edit the height defined as basic row
* set all the grid as static

## Usage

All questions about the grid generation and the possible configurable options are presents on [react-grid-layout repository](https://github.com/STRML/react-grid-layout). I trust you know how to use react-grid-layout and will explain now how React-Grid-Layout-builder works. React-Grid-Layout-builder is used as a wrapper on an existing React-Grid-Layout component. To demonstrate how to us the builder, I'll try it on the [responsive react-grid-layout example](https://github.com/STRML/react-grid-layout#providing-grid-width)

To have the builder work with the [responsive react-grid-layout example](https://github.com/STRML/react-grid-layout#providing-grid-width), we must:

* import the connect function `connectReactGridLayoutBuilder` to track any change in the `react-grid-layout` component
* import the builder `ReactGridLayoutBuilder` to offer additional field to modify the react-grid-layout configuration
* put the react-grid-layout configuration into a state to be able to modify it from a callback
* sent the callback that modify the react-grid-layout configuration to both `connectReactGridLayoutBuilder` and `ReactGridLayoutBuilder`

### An example:

```javascript
import {Responsive, WidthProvider} from 'react-grid-layout';
import ReactGridLayoutBuilder, {connectReactGridLayoutBuilder} from 'react-grid-layout-builder';

//the responsive react-grid-layout is wrapped to be connected to the builder functions
const ResponsiveReactGridLayout = connectReactGridLayoutBuilder(WidthProvider(Responsive)); 

...
constructor() {
  super();
  //the grid layout is put on the state
  this.state.gridLayout = 
    layouts: getLayoutsFromSomewhere(),
    breakpoints: {lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0},
    cols: {lg: 12, md: 10, sm: 6, xs: 4, xxs: 2}
  }
}

//any update from the grid layout or from the builder will update the component state to update both ReactGridLayoutBuilder & ResponsiveReactGridLayout data
updateConfig = (newGridLayout) => {
  this.setState({gridLayout: newGridLayout});
}
  
//An additional props updateConfigFunc must be provided to ResponsiveReactGridLayout to be able to track relevant changes when playing with the grid layout component it-self
//The usual react grid layout configuration, put in the component state, is sent to the grid layout component.
render() {
  return (
    <div>
      <ReactGridLayoutBuilder reactGridLayout={this.props.conf} updateConfigFunc={this.props.updateConfig}/>
      <ResponsiveReactGridLayout className="layout" 
        {...this.state.gridLayout}
        updateConfigFunc={this.props.updateConfig}>
        <div key={"1"}>1</div>
        <div key={"2"}>2</div>
        <div key={"3"}>3</div>
      </ResponsiveReactGridLayout>
    </div>
  )
}
...
```

### API:

#### ReactGridLayoutBuilder
`<ReactGridLayoutBuilder/>` is a react component that renders fields to edit some of the [react-grid-layout possible props](https://github.com/STRML/react-grid-layout#grid-layout-props).

This component uses [Bootstrap](http://getbootstrap.com/getting-started/) & [React-Bootstrap](https://react-bootstrap.github.io/) for rendering the fields

The builder can be used to set:

```javascript
// If true, the container height swells and contracts to fit contents
autoSize: ?boolean = true,

// {name: pxVal}, e.g. {lg: 1200, md: 996, sm: 768, xs: 480}
// Breakpoint names are arbitrary but must match in the cols and layouts objects.
breakpoints: ?Object = {lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0},

// # of cols. This is a breakpoint -> cols map, e.g. {lg: 12, md: 10, ...}
cols: ?Object = {lg: 12, md: 10, sm: 6, xs: 4, xxs: 2},

// If true, the layout will compact vertically
verticalCompact: ?boolean = true,

// Margin between items [x, y] in px.
margin: ?[number, number] = [10, 10],

// Rows have a static height, but you can change this based on breakpoints
// if you like.
rowHeight: ?number = 150,

//
// Flags
//
isDraggable: ?boolean = true,
isResizable: ?boolean = true,
```


#### connectReactGridLayoutBuilder

`connectReactGridLayoutBuilder` is a high order component wrapping the responsive react-grid-layout component to trigger a callback `updateConfig` when the layout is changed there. the callback has as parameter the [react-grid-layout possible props](https://github.com/STRML/react-grid-layout#grid-layout-props). 

```javascript
import {Responsive, WidthProvider} from 'react-grid-layout';
import {connectReactGridLayoutBuilder} from 'react-grid-layout-builder';

//the responsive react-grid-layout is wrapped to be connected to the builder functions
const ResponsiveReactGridLayout = connectReactGridLayoutBuilder(WidthProvider(Responsive)); 

...
constructor() {
  super();
  //the grid layout is put on the state
  this.state.gridLayout = 
    layouts: getLayoutsFromSomewhere(),
    breakpoints: {lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0},
    cols: {lg: 12, md: 10, sm: 6, xs: 4, xxs: 2}
  }
}

//any update from the grid layout or from the builder will update the component state to update both ReactGridLayoutBuilder & ResponsiveReactGridLayout data
updateConfig = (newGridLayout) => {
  this.setState({gridLayout: newGridLayout});
}
  
//An additional props updateConfigFunc must be provided to ResponsiveReactGridLayout to be able to track relevant changes when playing with the grid layout component it-self
//The usual react grid layout configuration, put in the component state, is sent to the grid layout component.
render() {
  return (
    <div>
      <ResponsiveReactGridLayout className="layout" 
        {...this.state.gridLayout}
        updateConfigFunc={this.props.updateConfig}>
        <div key={"1"}>1</div>
        <div key={"2"}>2</div>
        <div key={"3"}>3</div>
      </ResponsiveReactGridLayout>
    </div>
  )
}
...
```

#### withOpeningDock

`withOpeningDock` is a high order component used to have a react component displayed only if the user choose it. This is a simple UI component and you can choose to not use it if you want.

If you want to use it, wrap `ReactGridLayoutBuilder` with it:

```javascript
import ReactGridLayoutBuilder, {withOpeningDock} from '../../src';
const ReactGridLayoutBuilder = connectReactGridLayoutBuilder(WidthProvider(Responsive)); 
var DockedReactGridLayoutBuilder = withOpeningDock(ReactGridLayoutBuilder);

...
render() {
  return (
    <div>
      <DockedReactGridLayoutBuilder reactGridLayout={this.props.conf} updateConfigFunc={this.props.updateConfig}/>
      <ResponsiveReactGridLayout className="layout" 
        {...this.state.gridLayout}
        updateConfigFunc={this.props.updateConfig}>
        <div key={"1"}>1</div>
        <div key={"2"}>2</div>
        <div key={"3"}>3</div>
      </ResponsiveReactGridLayout>
    </div>
  )
}
...
```

## TODO List

- [ ] Get feedbacks and improve
- [ ] Transform the generated layout as static
- [ ] Choose which item is static or not
- [ ] Choose which item is displayed or not
- [ ] Provide an editor for not-responsive grid layout
- [x] Provide an editor based on material UI (thanks to [@nywooz](https://github.com/nywooz))
- [ ] Have the materia-ui demo working without bootstrap css (fix css grid)
- [x] Save & load the current custom layout in local storage (thanks to [@nywooz](https://github.com/nywooz))
- [x] Download the current layout & upload a custom one (thanks to [@nywooz](https://github.com/nywooz))


CHANGELOG.md
# Changelog

## [Unreleased] (need to update branch gh-page when merged on "master")

### Added
Transform the generated layout as static

### Added
Choose which item is static or not

## [3.0.0] - 03/21/2019

### Added
Download the current layout & upload a custom one

### Added
"Save or load a custom layout" using local storage as one from the TODO List

## [2.0.0] - 04/02/2018

### Added

new version of this editor with Material-ui (thank to [@nywooz](https://github.com/nywooz) for contributing):

```javascript
import { MaterialUIReactGridLayoutBuilder } from 'react-grid-layout-builder';
```

Caveat: 

The material ui editor requires the bootstrap css for now for the grid layout. This will be fixed in future versions. (will required the latest version fo material-ui)


### Changed

#### default export removed

This package doesn't export the editor as default any more. The default export is replaced by two simple export. To migrate, replace 

```javascript
import ReactGridLayoutBuilder from 'react-grid-layout-builder';
```

with:

```javascript
import { BootstrapReactGridLayoutBuilder as ReactGridLayoutBuilder } from 'react-grid-layout-builder';
```

#### New Demo/dev Url for material-ui or bootstrap:

In dev mode, [http://localhost:3100/](http://localhost:3100/) is replaced by:

- [http://localhost:3100/bootstrap/index.html](http://localhost:3100/bootstrap/index.html) for bootstrap editor
- [http://localhost:3100/material-ui/index.html](http://localhost:3100/material-ui/index.html) for material-ui editor


## [1.0.0]

### Changed

- Update to `react` 16
- remove lib from git repo



demo/src/bootstrap/app.js
import React from "react";
import { Responsive } from "react-grid-layout";
import {
  BootstrapReactGridLayoutBuilder,
  connectReactGridLayoutBuilder,
  withOpeningDock
} from "../../../src";
import { Grid, Row, Col, Button } from "react-bootstrap";

const WidthProvider = require("react-grid-layout").WidthProvider;
const ResponsiveReactGridLayout = connectReactGridLayoutBuilder(
  WidthProvider(Responsive)
);
const DockedReactGridLayoutBuilder = withOpeningDock(
  BootstrapReactGridLayoutBuilder
);

function generateLayout() {
  return _.map(_.range(0, 25), function(item, i) {
    const y = Math.ceil(Math.random() * 4) + 1;
    return {
      x: (_.random(0, 5) * 2) % 12,
      y: Math.floor(i / 6) * y,
      w: 2,
      h: y,
      i: i.toString(),
      static: Math.random() < 0.05
    };
  });
}

const defaultReactGridLayoutProps = {
  layouts: {
    lg: getFromLS("layouts", "lg") || generateLayout()
  },
  rowHeight: 30,
  breakpoints: { lg: 1200, md: 996, sm: 768, xs: 480, xxs: 0 },
  cols: { lg: 12, md: 10, sm: 6, xs: 4, xxs: 2 }
};

class ReactGridLayoutBuilderDemo extends React.Component {
  generateDOM = () => {
    const isAllStatic = this.props.conf.isAllStatic;
    const toggleStatic = this.props.toggleStatic;
    return _.map(this.props.conf.layouts.lg, function(l, i) {
      const staticGrid = isAllStatic ? true : l.static;
      return (
        <div key={i} className={staticGrid ? "static" : ""}>
          <div>
            <i
              className={`float-right p-1 fa fa-thumbtack ${
                staticGrid ? "" : "fa-rotate-90"
              }`}
              aria-hidden="true"
              onClick={() => toggleStatic(l)}
            />
          </div>
          {staticGrid ? (
            <span
              className="text"
              title="This item is static and cannot be removed or resized."
            >
              Static - {i}
            </span>
          ) : (
            <span className="text">{i}</span>
          )}
        </div>
      );
    });
  };

  onLayoutChange = layout => {
    saveToLS("layouts", layout, "lg");
  };

  render() {
    return (
      <div>
        <DockedReactGridLayoutBuilder
          reactGridLayout={this.props.conf}
          updateConfigFunc={this.props.updateConfig}
        />
        <ResponsiveReactGridLayout
          onLayoutChange={this.onLayoutChange}
          {...this.props.conf}
          updateConfigFunc={this.props.updateConfig}
        >
          {this.generateDOM()}
        </ResponsiveReactGridLayout>
      </div>
    );
  }
}

export default class App extends React.Component {
  constructor() {
    super();
    this.state = defaultReactGridLayoutProps;
  }
  updateConfig = config => {
    this.setState(config);
  };

  toggleStatic = grid => {
    const newlayouts = { ...this.state.layouts };
    const newitems = newlayouts.lg.slice();
    const index = newitems.findIndex(function(item) {
      return item.i === grid.i;
    });
    if (index === -1) return;

    newitems[index].static = !grid.static;

    this.setState({
      layouts: newlayouts
    });
  };

  handleFile = element => {
    // https://www.html5rocks.com/en/tutorials/file/dndfiles/
    const files = element.files;
    if (!files.length) {
      alert("Please select a file!");
      return;
    }
    const file = files[0];
    const start = 0;
    const stop = file.size - 1;

    const reader = new FileReader();
    // If we use onloadend, we need to check the readyState.
    reader.onloadend = evt => {
      if (evt.target.readyState == FileReader.DONE) {
        // DONE == 2
        const layouts = { ...this.state.layouts };
        layouts.lg = JSON.parse(evt.target.result).lg;
        this.setState({ layouts });
      }
    };

    const blob = file.slice(start, stop + 1);
    reader.readAsBinaryString(blob);
  };

  render() {
    return (
      <Grid fluid>
        <Row>
          <Col xs={12} md={9} lg={10}>
            <ReactGridLayoutBuilderDemo
              conf={this.state}
              updateConfig={this.updateConfig}
              toggleStatic={this.toggleStatic}
            />
          </Col>
          <Col xs={12} md={3} lg={2}>
            <h1>React Grid Layout Builder</h1>
            <h3>A user interface to build react-grid-layout</h3>
            <a
              href="https://github.com/damienleroux/react-grid-layout-builder"
              title="View on github"
              target="blank"
            >
              <i className="fa fa-github" aria-hidden="true" /> View on github
            </a>
            <h4>
              powered by
              <a
                href="https://github.com/STRML/react-grid-layout"
                target="blank"
              >
                @react-grid-layout
              </a>
            </h4>
            <h4>
              powered by
              <a href="https://react-bootstrap.github.io/" target="blank">
                @react-bootstrap
              </a>
            </h4>
            <div className="input-group mb-3">
              <div className="custom-file">
                <input
                  onChange={e => this.handleFile(e.target)}
                  accept=".json"
                  type="file"
                  className="custom-file-input"
                  id="inputGroupFile01"
                />
                <label
                  className="custom-file-label text-truncate"
                  htmlFor="inputGroupFile01"
                  aria-describedby="inputGroupFileAddon01"
                >
                  Choose file
                </label>
              </div>
            </div>

            <a
              href={`data:application/json;charset=utf-8,${encodeURIComponent(
                JSON.stringify(this.state.layouts)
              )}`}
              download={`layout ${new Date().toISOString()}.json`}
              className="btn btn-primary mb-3"
            >
              Download layout
            </a>
            <h5>Generated Layout:</h5>
            <pre>
              <code>{JSON.stringify(this.state, null, 2)}</code>
            </pre>
          </Col>
        </Row>
      </Grid>
    );
  }
}

function getFromLS(prop, key) {
  let ls = {};
  if (global.localStorage) {
    try {
      ls = JSON.parse(global.localStorage.getItem(key)) || {};
    } catch (e) {
      /*Ignore*/
    }
  }
  //return [];
  return ls[prop];
}

function saveToLS(valueKey, value, key) {
  if (global.localStorage) {
    global.localStorage.setItem(
      key,
      JSON.stringify({
        [valueKey]: value
      })
    );
  }
}
